MMCore/MMCore.cpp

```diff
@@ -2658,6 +2691,203 @@ bool CMMCore::getShutterOpen() throw (CMMError)
+                  unsigned byteDepth, unsigned nComponents, bool addLegacyMetadata) throw (CMMError)
+{    
+    // Essential metadata for interpreting the image: Width, height, and pixel type
+    md.PutImageTag(MM::g_Keyword_Metadata_Width, (unsigned int) width);
+    md.PutImageTag(MM::g_Keyword_Metadata_Height, (unsigned int) height);
+    
+    if (byteDepth == 1)
+         md.PutImageTag(MM::g_Keyword_PixelType, MM::g_Keyword_PixelType_GRAY8);
+    else if (byteDepth == 2)
+         md.PutImageTag(MM::g_Keyword_PixelType, MM::g_Keyword_PixelType_GRAY16);
+    else if (byteDepth == 4) {
+         if (nComponents == 1)
+            md.PutImageTag(MM::g_Keyword_PixelType, MM::g_Keyword_PixelType_GRAY32);
+         else
+            md.PutImageTag(MM::g_Keyword_PixelType, MM::g_Keyword_PixelType_RGB32);
+    } else if (byteDepth == 8)
+         md.PutImageTag(MM::g_Keyword_PixelType, MM::g_Keyword_PixelType_RGB64);
+    else
+         md.PutImageTag(MM::g_Keyword_PixelType, MM::g_Keyword_PixelType_Unknown);
+
+   // Needed for acquisition engine
+   if (!md.HasTag(MM::g_Keyword_Metadata_CameraLabel)) {
+     md.put(MM::g_Keyword_Metadata_CameraLabel, pCam->GetLabel());
+   }
+   
+   // Return if no additional metadata requested
+   if (metadataProfileFlag_ == 0) {
+      return;
+   }
+
+   // Group 2: Bit depth
+   if (metadataProfileFlag_ >= 2) {
+      md.put(MM::g_Keyword_Metadata_BitDepth, CDeviceUtils::ConvertToString((long) pCam->GetBitDepth()));
+   }
+   
+   // Group 3: Camera settings (ROI, binning)
+   if (metadataProfileFlag_ >= 3) {
+      unsigned x, y, xSize, ySize;
+      pCam->GetROI(x, y, xSize, ySize);
+      std::string roiTag = std::to_string(x) + "-" + std::to_string(y) + "-" + 
+                           std::to_string(xSize) + "-" + std::to_string(ySize);
+      md.put("ROI", roiTag);
+      
+      try {
+         std::string binning = pCam->GetProperty("Binning");
+         md.put("Binning", binning);
+      }
+      catch (const CMMError&) { }
+   }
+   
+   // Group 4: Timing metadata
+   if (metadataProfileFlag_ >= 4) {
+      std::lock_guard<std::mutex> lock(imageNumbersMutex_);
+      std::string cameraName = md.GetSingleTag(MM::g_Keyword_Metadata_CameraLabel).GetValue();
+      if (imageNumbers_.find(cameraName) == imageNumbers_.end()) {
+         imageNumbers_[cameraName] = 0;
+      }
+      md.put(MM::g_Keyword_Metadata_ImageNumber, CDeviceUtils::ConvertToString(imageNumbers_[cameraName]));
+      ++imageNumbers_[cameraName];
+      
+      if (!md.HasTag(MM::g_Keyword_Elapsed_Time_ms)) {
+         using namespace std::chrono;
+         auto elapsed = steady_clock::now() - startTime_;
+         md.put(MM::g_Keyword_Elapsed_Time_ms,
+                           std::to_string(duration_cast<milliseconds>(elapsed).count()));
+      }
+      
+      auto now = std::chrono::system_clock::now();
+      md.put(MM::g_Keyword_Metadata_TimeInCore, FormatLocalTime(now));
+   }
```

Comment:
having these keys guarded by a non-default metadata flag level is a breaking change (that showed up in pymmcore-plus tests).    I do like the flags to guard various metadata-retrieval, but if we don't want to provide these by default, it first needs to be deprecated.   specifically:
- g_Keyword_Metadata_ImageNumber
- g_Keyword_Elapsed_Time_ms
- g_Keyword_Metadata_TimeInCore

were all present in metadata previously